Fix issue 12954 - Deprecated only works with string literals

Geod24 opened the pull request:
https://issues.dlang.org/show_bug.cgi?id=12954

Allow code to use `enum` and `static {immutable,const}` for the deprecation message.
This will allow for more dynamic deprecation message (e.g. with release informations).
It purposedly doesn't support direct function call to get the deprecation message.

Geod24 committed:
Fix issue 12954 - Deprecated only works with string literals

Allow code to use `enum` and `static {immutable,const}` for the deprecation message.
This will allow for more dynamic deprecation message (e.g. with release informations).
It purposedly doesn't support direct function call to get the deprecation message.

yebblies commented on src/attrib.d:
```diff
@@ -411,6 +411,11 @@ public:
         assert(msg);
         char* depmsg = null;
+        sc = sc.startCTFE();
+        this.msg = this.msg.semantic(sc);
```
Prefixing with `this` is not necessary.

9rnsr commented on src/attrib.d:
```diff
@@ -411,6 +411,11 @@ public:
+        this.msg = this.msg.semantic(sc);
+        this.msg = resolveProperties(sc, this.msg);
+        sc = sc.endCTFE();
+        this.msg = this.msg.optimize(WANTexpand);
```
To evaluate the `msg` expression, `setScope` is too early. For example, it doesn't work properly for:

```d
enum a1 = "a", b1 = "b";
deprecated(a1 ~ b1) int v1 = 1; // OK

deprecated(a2 ~ b2) int v2 = 2; // NG
enum a2 = "a", b2 = "b";
```

UDA doesn't have same issue. In order to avoid forward reference issues, UDA is evaluated at when `__traits(getAttribute)` is requested (it happens in `semantic` stage at the earliest), or done in `UserAttributeDeclaration.semantic2` stage at the latest.

Geod24 replied:
Thanks for the tip.
So it either have to be part of semantic (X) or `scope` / `dsymbol` will need to be changed store the `DeprecatedDeclaration` instead of just the message (as it's done with UDA currently) ? I guess the later would be cleaner (albeit more invasive) ?

Geod24 commented:
Updated

yebblies commented on src/attrib.d:
```diff
@@ -409,17 +410,27 @@ public:
+    override void semantic2(Scope* sc)
+    {
         assert(msg);
         char* depmsg = null;
```
Dead code?

Geod24 replied:
RIchtig ! Fixed.

yebblies commented:
This looks ok to me.  @9rnsr ?

9rnsr commented on src/attrib.d:
```diff
@@ -409,17 +410,26 @@ public:
+        msg = msg.semantic(sc);
+        msg = resolveProperties(sc, msg);
+        sc = sc.endCTFE();
+        msg = msg.optimize(WANTexpand);
```
Need to use `msg.ctfeInterpret()` instead.

9rnsr commented on src/attrib.d:
```diff
@@ -394,6 +394,7 @@ extern (C++) final class DeprecatedDeclaration : StorageClassDeclaration
 {
 public:
     Expression msg;
+    const(char)* message;
```
`msg` vs `message` is a little confusing. I want to suggest either:
1. rename original `msg` field to `emsg`
2. rename new `message` field to `msgstr`
or something better.

ibuclaw replied:
Original code used ``depmsg``.  Another alternative to that would be ``depstr``

9rnsr commented on src/attrib.d:
```diff
@@ -409,17 +410,26 @@ public:
+        scx.depdecl = this;
+        StorageClassDeclaration.setScope(scx);
+        scx.pop();
+    }
```
When you save a scope `scx` by `setScope(scx)`, the scx should be an *enclosing scope* of this `DeprecatedDeclaration`, ant it should be equivalent with the scope object which will be passed to `semantic(Scope* sc)`, `semantic2(Scope* sc)`, and others.

However, this code saves a new scope which is set the `depdecl` field to `this`. It will mean the enclosing deprecation attribute of `this` is `this` - there's recursive structure! Of course it's not good.

Instead, you can override `newScope` member function to construct a new scope for the members of `DeprecatedDeclaration`.

9rnsr commented on src/dsymbol.d:
```diff
@@ -191,7 +191,7 @@ public:
     PASS semanticRun;
 
-    char* depmsg;           // customized deprecation message
+    DeprecatedDeclaration depdecl;           // customized deprecation message
```
Adding a new `decdecl` field similar to `userAttribDecl` is right direction.

Geod24 commented:
Updated according to @9rnsr 's comments:
- `s/message/msgstr/`
- `msg.optimize(WANTexpand)` -> `msg.ctfeInterpret()`
- Changed `setScope` for `newScope`.

Also documented the functions and added a small test with a deprecated class + deprecated member to reflect the new implementation.

@9rnsr : `StorageClassDeclaration.newScope` was `final`, so I had to remove this annotation. Not sure how good that is though.

9rnsr commented:
Excellent work!

> StorageClassDeclaration.newScope was final, so I had to remove this annotation. Not sure how good that is though.

Of course there's no problem. The `final` was there just because there was no derived class who overrides `StorageClassDeclaration.newScope` method. Removing it is legitimate change.

9rnsr commented:
I'll get this enhancement change. @Geod24, can you also write a changelog section for the "Language Enhancement"? In #5425, I'm resetting `changelog.dd`. After it's merged, let's rebase this PR on master, with the new changelog text.

Geod24 commented:
> @Geod24, can you also write a changelog section for the "Language Enhancement"?

Will do once #5425 is merged. I'll also prepare a P.R. to dlang.org documenting the grammar change.

9rnsr commented:
#5425 is merged.

Geod24 commented:
Rebased on master, changelog entry added.

9rnsr commented:
Thanks!

9rnsr commented:
Auto-merge toggled on

Geod24 commented:
Dlang.org P.R.: https://github.com/D-Programming-Language/dlang.org/pull/1229

9rnsr commented:
Mm, I had overlooked this statement written in the summary.

> It purposedly doesn't support direct function call to get the deprecation message.

While the reviewing I've proposed this change.

> `msg.optimize(WANTexpand)` -> `msg.ctfeInterpret()`

By that, now any compile time function calls for the custom deprecation message is properly allowed...

@Geod24 , why you've thought that function call use has to be rejected as a custom message? I'd just hear your argument.

Geod24 commented:
> @Geod24 , why you've thought that function call use has to be rejected as a custom message? I'd just hear your argument.

I'm very happy with it accepting direct function call. I think it was originally rejected because I was afraid it would cause forward references, but that was down to using `setScope` as you pointed.
Will update the dlang.org P.R. accordingly.

9rnsr commented:
I understand, thanks.

MartinNowak commented:
Seems like this introduced a small regression.
[Issue 15815 – [Reg 2.071-devel] deprecation for aliased template instance not shown](https://issues.dlang.org/show_bug.cgi?id=15815)
